pkg/storage/s3/s3.go

```diff
@@ -66,11 +69,37 @@ func NewDriver(crname string, crnamespace string, c *opapi.ImageRegistryConfigSt
 	}
 }
 
-// GetType returns the name of the storage driver
+// NewDriverFromResource creates a new driver from an existing image registry resource
+func NewDriverFromResource(cr *opapi.ImageRegistry) (*driver, error) {
+	if cr.Status.Storage.State.S3 == nil {
```

Comment:
why is this looking at status instead of spec?

I would expect the following sequence of events during bootstrap:

1) we identify we want to use s3 storage
2) we invoke NewDriver to create an s3 driver
3) we use the driver to populate/initialize spec.storage.s3 in the CR we are bootstrapping
4) we create the CR
5) the controller sees the CR, the driver gets invoked to turn the requested "spec.storage.s3" into a storage that exists
6) the driver, after creating the storage, updates status.storage.s3 with the details of the storage it created, and updates the condition storagexists to true

And of course in cases where the user creates a CR themselves, the flow would be:
1) the controller sees the CR, the driver gets invoked to turn the requested "spec.storage.s3" into a storage that exists
2) the driver, after creating the storage, updates status.storage.s3 with the details of the storage it created, and updates the condition storagexists to true

Reply:
If the current status.storage is GCS, but the spec.storage is S3, then what's going to happen when you invoke this function with the CR?  How will the operator move forward to transition onto s3 storage?

The one case where you're doing this that makes sense is the finalizer.... when tearing down, you want to tear down the current state, not the requested state.  But in the generator it does not make sense.  

What you should probably do is just pass a "Storage" object to NewDriverFromStorage().  Then depending on context you can invoke NewDriveFromStorage(spec.Storage) or NewDriverFromStorage(status.Storage).

And CreateStorage/etc should just look at the storage object that was passed to NewDriverFromStorage, rather than being hardcoded to look at spec.

Reply:
ok, this should be sorted out now, thanks!